tests/e2e/test_parameterized_queries.py

```diff
@@ -0,0 +1,382 @@
+    for @both_approaches.
+    """
+
+    NATIVE_QUERY = "SELECT :p AS col"
+
+    inline_type_map = {
+        int: "int_col",
+        float: "float_col",
+        Decimal: "decimal_col",
+        str: "string_col",
+        bool: "boolean_col",
+        datetime.date: "date_col",
+        datetime.datetime: "timestamp_col",
+        type(None): "null_col",
+    }
+
+    @pytest.fixture(scope="class")
+    def inline_table(self):
+        """This table is necessary to verify that a parameter sent with INLINE
+        approach can actually write to its analogous data type.
+
+        For example, a Python Decimal(), when rendered inline, should be able
+        to read/write into a DECIMAL column in Databricks
+
+        Note that this fixture doesn't clean itself up. So the table will remain
+        in the schema for use by subsequent test runs.
+        """
+
+        query = """
+            CREATE TABLE IF NOT EXISTS pysql_e2e_inline_param_test_table (
+            null_col INT,
+            int_col INT,
+            float_col FLOAT,
+            decimal_col DECIMAL(10, 2),
+            string_col STRING,
+            boolean_col BOOLEAN,
+            date_col DATE,
+            timestamp_col TIMESTAMP
+            ) USING DELTA
+        """
+
+        with self.connection() as conn:
+            with conn.cursor() as cursor:
+                cursor.execute(query)
+
+    def compute_says_it_doesnt_support_native_params(self):
+        """we test against dogfood. dogfood exposes a protocol version that suggests it can't handle
+        native queries, even though it does. for test coverage, we can mock the local protocol check.
+        but we want to remove this mock once dogfood properly advertises its version.
+        """
+
+        with self.connection() as conn:
+            return (
+                conn.protocol_version
+                < ttypes.TProtocolVersion.SPARK_CLI_SERVICE_PROTOCOL_V8
+            )
+
+    @contextmanager
+    def conditional_protocol_patch(self, bypass=False):
+        """This fixture will be removed once dogfood advertises its protocol version correctly.
+        
+        Note that there is an equivalent patch in sqlalchemy/test/test_suite.py which should be 
+        removed at the same time as this one. That one is encapsulated in a function called
+        start_protocol_patch()"""
+
+        if bypass:
+            yield None
+
+        if not self.compute_says_it_doesnt_support_native_params():
+            yield None
+
+        with patch(
+            "databricks.sql.client.Connection.server_parameterized_queries_enabled",
+            return_value=True,
+        ) as mock_parameterized_queries_enabled:
+            try:
+                yield mock_parameterized_queries_enabled
+            finally:
+                pass
+
+    def _inline_roundtrip(self, params: dict):
+        target_column = self.inline_type_map[type(params.get("p"))]
```

Comment:
This INSERT, SELECT, DELETE dance is necessary because simply selecting "SELECT %(param)s" in INLINE mode would always return a `str` and the nature of the test is to confirm that types are maintained.

Reply:
I'll add the above context to the docstring for this method.

Reply:
Done.